language/move-lang/tests/move_check_testsuite.rs

```diff
@@ -42,19 +48,81 @@ fn format_diff(expected: String, actual: String) -> String {
     ret
 }
 
-// Runs all tests under the test/testsuite directory.
 fn move_check_testsuite(path: &Path) -> datatest_stable::Result<()> {
-    let targets: Vec<String> = vec![path.to_str().unwrap().to_owned()];
-    let deps = move_stdlib::move_stdlib_files();
+    // A test is marked that it should also be compiled in test mode by having a `path.unit_test`
+    // file.
+    if path.with_extension(TEST_EXT).exists() {
+        let test_exp_path = format!(
+            "{}.unit_test.{}",
+            path.with_extension("").to_string_lossy(),
+            EXP_EXT
+        );
+        let test_out_path = format!(
+            "{}.unit_test.{}",
+            path.with_extension("").to_string_lossy(),
+            OUT_EXT
+        );
+        run_test(
+            path,
+            &Path::new(&test_exp_path),
+            &Path::new(&test_out_path),
+            Flags::testing(),
+        )?;
+    }
 
     let exp_path = path.with_extension(EXP_EXT);
     let out_path = path.with_extension(OUT_EXT);
+    run_test(path, &exp_path, &out_path, Flags::empty())
+}
+
+// Runs all tests under the test/testsuite directory.
+fn run_test(
+    path: &Path,
+    exp_path: &Path,
+    out_path: &Path,
+    flags: Flags,
+) -> datatest_stable::Result<()> {
+    let targets: Vec<String> = vec![path.to_str().unwrap().to_owned()];
+    let deps = move_stdlib::move_stdlib_files();
+    let mut compilation_env = CompilationEnv::new(flags);
+
+    let (files, pprog_and_comments_res) = move_parse(&targets, &deps, None, false)?;
+    let errors = pprog_and_comments_res
+        .and_then(|(_, pprog)| {
```

Comment:
the amount of and_thens is a bit concerning. Maybe breaking this into a helper with ? might be better?